File: smart-quiz-hub-frontend/src/app/core/services/review.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import {
  ApiResponse, PagedResponse, McqResponse, ReviewRequest,
  AssignReviewerRequest, BulkAssignRequest, BulkAssignResponse
} from '../models';
import { environment } from '../../../environments/environment';

/** Inline request body for POST /reviews/questions/bulk-decision (admin only). */
export interface BulkDecisionRequest {
  questionIds: number[];
  /** APPROVED | REJECTED | MODIFICATION_REQUESTED */
  decision: string;
  /** Required when decision is REJECTED or MODIFICATION_REQUESTED. */
  comments?: string;
}

/** Inline response shape for POST /reviews/questions/bulk-decision. */
export interface BulkDecisionResponse {
  processed: number;
  skipped: number;
  skippedReasons: string[];
}

@Injectable({ providedIn: 'root' })
export class ReviewService {
  private base = `${environment.apiUrl}/reviews`;

  private http = inject(HttpClient);

  assignReviewer(questionId: number, req: AssignReviewerRequest): Observable<ApiResponse<McqResponse>> {
    return this.http.post<ApiResponse<McqResponse>>(
      `${this.base}/questions/${questionId}/assign`, req);
  }

  bulkAssignReviewer(req: BulkAssignRequest): Observable<ApiResponse<BulkAssignResponse>> {
    return this.http.post<ApiResponse<BulkAssignResponse>>(
      `${this.base}/questions/bulk-assign`, req);
  }

  submitReview(questionId: number, req: ReviewRequest): Observable<ApiResponse<McqResponse>> {
    return this.http.post<ApiResponse<McqResponse>>(
      `${this.base}/questions/${questionId}/decision`, req);
  }

  bulkDecision(questionIds: number[], decision: string, comments?: string):
      Observable<ApiResponse<BulkDecisionResponse>> {
    const req: BulkDecisionRequest = { questionIds, decision, comments };
    return this.http.post<ApiResponse<BulkDecisionResponse>>(
      `${this.base}/questions/bulk-decision`, req);
  }

  getPendingReviews(page = 0, size = 10): Observable<ApiResponse<PagedResponse<McqResponse>>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<ApiResponse<PagedResponse<McqResponse>>>(`${this.base}/pending`, { params });
  }

  getReadyForReview(page = 0, size = 10): Observable<ApiResponse<PagedResponse<McqResponse>>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<ApiResponse<PagedResponse<McqResponse>>>(`${this.base}/ready`, { params });
  }

  getReviewedByMe(page = 0, size = 10): Observable<ApiResponse<PagedResponse<McqResponse>>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<ApiResponse<PagedResponse<McqResponse>>>(`${this.base}/reviewed`, { params });
  }

  autoAssignReviewers(): Observable<ApiResponse<null>> {
    return this.http.post<ApiResponse<null>>(`${this.base}/auto-assign`, {});
  }
}

File: smart-quiz-hub-backend/src/main/java/com/accenture/smartquiz/service/ReviewService.java
package com.accenture.smartquiz.service;

import com.accenture.smartquiz.dto.request.AssignReviewerRequest;
import com.accenture.smartquiz.dto.request.BulkAssignRequest;
import com.accenture.smartquiz.dto.request.BulkDecisionRequest;
import com.accenture.smartquiz.dto.request.ReviewRequest;
import com.accenture.smartquiz.dto.response.BulkAssignResponse;
import com.accenture.smartquiz.dto.response.BulkDecisionResponse;
import com.accenture.smartquiz.dto.response.McqResponse;
import com.accenture.smartquiz.dto.response.PagedResponse;
import com.accenture.smartquiz.entity.enums.McqStatus;
import com.accenture.smartquiz.security.SmartQuizUserDetails;
import org.springframework.data.domain.Pageable;

public interface ReviewService {

    McqResponse assignReviewer(Long questionId, AssignReviewerRequest request, SmartQuizUserDetails currentUser);

    BulkAssignResponse bulkAssignReviewer(BulkAssignRequest request, SmartQuizUserDetails currentUser);

    McqResponse startReview(Long questionId, SmartQuizUserDetails currentUser);

    McqResponse submitReview(Long questionId, ReviewRequest request, SmartQuizUserDetails currentUser);

    /** Admin-only: apply the same review decision to many questions, skipping per-item failures. */
    BulkDecisionResponse bulkDecision(BulkDecisionRequest request, SmartQuizUserDetails currentUser);

    PagedResponse<McqResponse> getPendingReviews(SmartQuizUserDetails currentUser, Pageable pageable);

    PagedResponse<McqResponse> getReadyForReview(SmartQuizUserDetails currentUser, Pageable pageable);

    /** Questions the current user has already decided (approved / rejected / sent back). */
    PagedResponse<McqResponse> getReviewedByMe(SmartQuizUserDetails currentUser, Pageable pageable);

    /** Auto-assign READY_FOR_REVIEW questions with no reviewer to the least-loaded eligible SME. */
    int autoAssignReviewers();
}
